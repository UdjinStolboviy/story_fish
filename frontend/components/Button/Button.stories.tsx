import { ComponentStoryObj, ComponentMeta } from "@storybook/react";
import { expect } from "@storybook/jest";
import { screen, userEvent } from "@storybook/testing-library";

import { Button } from "./Button";

export default {
  title: "Controls/Button",
  component: Button,
  args: {
    children: "Button",
  },
} as ComponentMeta<typeof Button>;

export const PrimaryButton: ComponentStoryObj<typeof Button> = {
  play: async ({ args }) => {
    await userEvent.click(screen.getByRole("button"));
    // @todo: https://github.com/storybookjs/storybook/issues/16941
    // await expect(args.onClick).toHaveBeenCalled();
  },
  args: {
    color: "primary",
  },
};

export const SecondaryButton: ComponentStoryObj<typeof Button> = {
  ...PrimaryButton,
  args: {
    color: "secondary",
  },
};

export const WarningButton: ComponentStoryObj<typeof Button> = {
  ...PrimaryButton,
  args: {
    color: "warning",
  },
};

export const DangerButton: ComponentStoryObj<typeof Button> = {
  ...PrimaryButton,
  args: {
    color: "danger",
  },
};
