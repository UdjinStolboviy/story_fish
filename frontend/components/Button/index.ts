export {
  Button,
  PrimaryButton,
  SecondaryButton,
  WarningButton,
  DangerButton,
} from "./Button";
