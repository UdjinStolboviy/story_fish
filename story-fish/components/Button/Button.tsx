import { MouseEvent } from "react";
import styled from "@emotion/styled";
import { css, SerializedStyles } from "@emotion/react";

export type Color =
  | "primary"
  | "secondary"
  | "tertiary"
  | "quaternary"
  | "danger"
  | "success"
  | "warning"
  | "info";

export type Props = {
  children: string;
  color?: Color;
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
};

export const getColor = (color?: Color): SerializedStyles => {
  switch (color) {
    case "primary":
      return css`
        background: #1ea7fd;
        color: #e4ebf5e6;
      `;
    case "secondary":
      return css`
        color: #5e5c64eb;
      `;
    case "tertiary":
      return css`
        background: #1ea7fd;
        color: #e4ebf5e6;
      `;
    case "quaternary":
      return css`
        background: #1ea7fd;
        color: #e4ebf5e6;
      `;
    case "danger":
      return css`
        background: #dc3545e6;
        color: #e4ebf5e6;
      `;
    case "success":
      return css`
        background: #1ea7fd;
        color: #e4ebf5e6;
      `;
    case "warning":
      return css`
        background: #ffca2ce6;
        color: #5e5c64eb;
      `;
    case "info":
      return css`
        background: #1ea7fd;
        color: #e4ebf5e6;
      `;
    default:
      return css``;
  }
};

export const Button = styled.button<Props>`
  all: unset;
  display: flex;
  align-items: center;
  justify-content: center;
  justify-self: center;
  user-select: none;
  cursor: pointer;
  font-size: 1.6rem;
  width: 15rem;
  height: 4rem;
  border-radius: 1rem;
  transition: all 0.4s ease;
  ${({ color }) => getColor(color)};
  box-shadow: 0.5vmin 0.5vmin 1vmin #c8d0e7, -0.5vmin -0.5vmin 1vmin #ffffff;
  &:hover {
    opacity: 0.8;
  }
  &:active {
    box-shadow: inset 0.5vmin 0.5vmin 1vmin #c8d0e7,
      inset -0.5vmin -0.5vmin 1vmin #ffffff;
  }
`;

Button.defaultProps = {
  color: "primary",
};
