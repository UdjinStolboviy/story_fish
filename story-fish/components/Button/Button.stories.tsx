import React from "react";
import { ComponentMeta, ComponentStoryObj } from "@storybook/react";

import { expect } from "@storybook/jest";

import { screen, userEvent } from "@storybook/testing-library";
import { Button } from "./Button";

export default {
  title: "Controls/Button",
  component: Button,
  args: { children: "Button" },
} as ComponentMeta<typeof Button>;

export const PrimaryButton: ComponentStoryObj<typeof Button> = {
  play: async ({ args }) => {
    await userEvent.click(screen.getByRole("button"));
    await expect(args.onClick).toHaveBeenCalled();
  },
  args: { color: "primary" },
};

export const SecondaryButton: ComponentStoryObj<typeof Button> = {
  ...PrimaryButton,
  args: { color: "secondary" },
};
