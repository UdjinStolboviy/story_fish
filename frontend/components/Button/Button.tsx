import { FC, MouseEvent } from "react";
import styled from "@emotion/styled";
import { css, SerializedStyles } from "@emotion/react";

import { AppTheme } from "@/styles/themes";
import { boxShadow, transition, borderRadius } from "@/components/styles";

export type Color = "primary" | "secondary" | "danger" | "warning";

export type Props = {
  /** Text in the button */
  children: string;
  /** Button color */
  color?: Color;
  /** Click handler */
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
};

export const getColors = (theme: AppTheme, color?: Color): SerializedStyles => {
  switch (color) {
    case "secondary":
      return css`
        color: ${theme.font.regular};
      `;
    case "primary":
    case "danger":
      return css`
        background: ${theme.components[color]};
        color: ${theme.font.button};
      `;
    case "warning":
      return css`
        background: ${theme.components[color]};
        color: ${theme.font.warning};
      `;
    default:
      return css``;
  }
};

export const Button = styled.button<Props>`
  all: unset;
  display: flex;
  justify-self: center;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  font-size: 1.6rem;
  width: 15rem;
  height: 4rem;
  ${borderRadius};
  ${({ theme, color }) => getColors(theme, color)};
  &:hover {
    opacity: 0.9;
  }
  ${transition()}
  ${({ theme }) =>
    boxShadow(theme.components.shadow1, theme.components.shadow2)}
  &:active {
    ${({ theme }) =>
      boxShadow(theme.components.shadow1, theme.components.shadow2, true)}
  }
`;

Button.defaultProps = {
  color: "primary",
};

type DefinedButton = Omit<Props, "color">;

export const PrimaryButton: FC<DefinedButton> = (props) => (
  <Button color="primary" {...props} />
);

export const SecondaryButton: FC<DefinedButton> = (props) => (
  <Button color="secondary" {...props} />
);

export const DangerButton: FC<DefinedButton> = (props) => (
  <Button color="danger" {...props} />
);

export const WarningButton: FC<DefinedButton> = (props) => (
  <Button color="warning" {...props} />
);
