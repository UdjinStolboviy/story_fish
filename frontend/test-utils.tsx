import { FC, ReactElement } from "react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@emotion/react";

import { rootReducer, RootState } from "./store";
import { Themes } from "./styles/themes";
import { Layout } from "./components/Layout";

type Options = { preloadedState?: RootState } & RenderOptions;

const customRender = (
  ui: ReactElement,
  { preloadedState, ...options }: Options = {}
) => {
  const store = configureStore({ reducer: rootReducer, preloadedState });

  const Wrapper: FC = ({ children }) => (
    <Provider store={store}>
      <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
    </Provider>
  );
  return render(ui, { wrapper: Wrapper, ...options });
};

const pageRender = (
  ui: ReactElement,
  { preloadedState, ...options }: Options = {}
) => {
  const store = configureStore({ reducer: rootReducer, preloadedState });

  const Wrapper: FC = ({ children }) => (
    <Provider store={store}>
      <Layout>{children}</Layout>
    </Provider>
  );
  return render(ui, { wrapper: Wrapper, ...options });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render, pageRender };
