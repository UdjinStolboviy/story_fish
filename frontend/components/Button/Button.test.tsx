import userEvent from "@testing-library/user-event";

import { render, screen } from "@/test-utils";

import { Button } from "./Button";

describe("Button test cases", () => {
  it("Render check", () => {
    const onClick = jest.fn();
    const { asFragment } = render(<Button onClick={onClick}>Button</Button>);

    expect(asFragment()).toMatchSnapshot();
  });
  it("Check onClick callback", () => {
    const onClick = jest.fn();

    render(<Button onClick={onClick}>Button</Button>);

    const element = screen.getByRole("button");

    userEvent.click(element);

    expect(onClick).toHaveBeenCalled();
  });
});
