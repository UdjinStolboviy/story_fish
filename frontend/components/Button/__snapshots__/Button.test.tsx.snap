// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button test cases Render check 1`] = `
<DocumentFragment>
  <button
    class="css-1rttvxo"
    color="primary"
  >
    Button
  </button>
</DocumentFragment>
`;
